### PR TITLE
fix for ghosted avatars thanks to check/add race

### PR DESCRIPTION
**Pull request by birarda**

### CODE REVIEW

File: interface/src/avatar/AvatarManager.cpp [hunk 1]
```diff
@@ -249,7 +256,6 @@ QVector<AvatarManager::LocalLight> AvatarManager::getLocalLights() const {
 }
 
 QVector<QUuid> AvatarManager::getAvatarIdentifiers() {
-    QReadLocker locker(&_hashLock);
```
Comment (birarda) [inline interface/src/avatar/AvatarManager.cpp]: I'm assuming that `_avatarHash.keys()` is safe to call without a read lock as it returns a `QList<QUuid>` of the keys at that point in time

**Comment by hifi-gustavo** [comment]
Build console output for commit: 40397add433f25c0dfa4baf39b711adf2dd673bc available at: https://gist.github.com/fa7d6bcdd13868cece34<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.dmg

File: interface/src/avatar/AvatarManager.cpp (hunk 1 above)
Comment (ZappoMan) [inline interface/src/avatar/AvatarManager.cpp]: I don't think that's a safe assumption... you need to make sure some other thread isn't in the midst of changing the hash. From the docs: `Note: All functions in this class are reentrant.`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 430cb7876dd99cfb7832620b2da65e40f0bf004f available at: https://gist.github.com/fa7d6bcdd13868cece34<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 22f5d4df6dece4e1c9f3c42d1cdced28e4b2dfa9 available at: https://gist.github.com/fa7d6bcdd13868cece34<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.dmg

File: interface/src/avatar/AvatarManager.cpp [hunk 2]
```diff
@@ -119,19 +119,19 @@ void AvatarManager::updateOtherAvatars(float deltaTime) {
     PerformanceTimer perfTimer("otherAvatars");
 
     // simulate avatars
-    AvatarHash::iterator avatarIterator = _avatarHash.begin();
-    while (avatarIterator != _avatarHash.end()) {
-        auto avatar = std::dynamic_pointer_cast<Avatar>(avatarIterator.value());
+    auto hashCopy = _avatarHash;
```
Comment (Atlante45) [inline interface/src/avatar/AvatarManager.cpp]: This needs a read locker

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 3]
```diff
@@ -22,13 +22,9 @@ AvatarHashMap::AvatarHashMap() {
...
 }
 
-void AvatarHashMap::withAvatarHash(std::function<void(const AvatarHash& hash)> callback) {
-    QReadLocker locker(&_hashLock);
-    callback(_avatarHash);
-}
 bool AvatarHashMap::isAvatarInRange(const glm::vec3& position, const float range) {
-    QReadLocker locker(&_hashLock);
-    foreach(const AvatarSharedPointer& sharedAvatar, _avatarHash) {
+    auto hashCopy = _avatarHash;
```
Comment (Atlante45) [inline libraries/avatars/src/AvatarHashMap.cpp]: Read lock

**Comment by hifi-gustavo** [comment]
Build console output for commit: 19d3d80ff97ec6e2aafff260fcc14b47482beaff available at: https://gist.github.com/fa7d6bcdd13868cece34<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b5bfd45bc4a822ff365d646b4d65c449655f649 available at: https://gist.github.com/fa7d6bcdd13868cece34<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6436.dmg

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 4]
```diff
@@ -44,16 +40,34 @@ AvatarSharedPointer AvatarHashMap::newSharedAvatar() {
 
 AvatarSharedPointer AvatarHashMap::addAvatar(const QUuid& sessionUUID, const QWeakPointer<Node>& mixerWeakPointer) {
     qCDebug(avatars) << "Adding avatar with sessionUUID " << sessionUUID << "to AvatarHashMap.";
-
-    AvatarSharedPointer avatar = newSharedAvatar();
+    
+    auto avatar = newSharedAvatar();
     avatar->setSessionUUID(sessionUUID);
     avatar->setOwningAvatarMixer(mixerWeakPointer);
-    QWriteLocker locker(&_hashLock);
```
Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: why are you deleting this? is the _hashLock being guarded somewhere else?

Comment (ZappoMan) [inline libraries/avatars/src/AvatarHashMap.cpp]: never mind I see it in the newOrExistingAvatar

**Pull request merged**